botorch/optim/optimize_mixed.py

```diff
@@ -148,26 +149,56 @@ def get_nearest_neighbors(
     Args:
         current_x: The design to find the neighbors of. A tensor of shape `d`.
         bounds: A `2 x d` tensor of lower and upper bounds for each column of `X`.
-        discrete_dims: A tensor of indices corresponding to binary and
-            integer parameters.
+        discrete_dims: A dictionary mapping indices of discrete dimensions
+            to a list of allowed values for that dimension.
 
     Returns:
         A tensor of shape `num_neighbors x d`, denoting all unique 1-Manhattan
         distance neighbors.
     """
+    # we first transform the discrete values in the current_x to its index in the
+    # values of the discrete_dims, so that we can easily increase/decrease
+    # the discrete dimensions by one.
+    # Map discrete values in current_x to their indices in discrete_dims
+    discrete_indices = []
+
+    t_discrete_dims = torch.tensor(
+        list(discrete_dims.keys()), dtype=torch.long, device=current_x.device
+    )
+
+    for dim, values in discrete_dims.items():
+        val = current_x[dim].item()
+        # Find the index of the closest value in values
+        idx = min(range(len(values)), key=lambda i: abs(values[i] - val))
+        discrete_indices.append(idx)
+    current_x_int = current_x.clone()
+    current_x_int[t_discrete_dims] = torch.tensor(
+        discrete_indices, dtype=current_x.dtype, device=current_x.device
+    )
     num_discrete = len(discrete_dims)
     diag_ones = torch.eye(num_discrete, dtype=current_x.dtype, device=current_x.device)
     # Neighbors obtained by increasing a discrete dimension by one.
-    plus_neighbors = current_x.repeat(num_discrete, 1)
-    plus_neighbors[:, discrete_dims] += diag_ones
+    plus_neighbors = current_x_int.repeat(num_discrete, 1)
+    plus_neighbors[:, t_discrete_dims] += diag_ones
     plus_neighbors.clamp_(max=bounds[1])
     # Neighbors obtained by decreasing a discrete dimension by one.
-    minus_neighbors = current_x.repeat(num_discrete, 1)
-    minus_neighbors[:, discrete_dims] -= diag_ones
+    minus_neighbors = current_x_int.repeat(num_discrete, 1)
+    minus_neighbors[:, t_discrete_dims] -= diag_ones
     minus_neighbors.clamp_(min=bounds[0])
     unique_neighbors = torch.cat([minus_neighbors, plus_neighbors], dim=0).unique(dim=0)
     # Also remove current_x if it is in unique_neighbors.
-    unique_neighbors = unique_neighbors[~(unique_neighbors == current_x).all(dim=-1)]
+    unique_neighbors = unique_neighbors[
+        ~(unique_neighbors == current_x_int).all(dim=-1)
+    ]
+
+    # Transform unique_neighbors back to the original non-integer space
+    for dim, values in discrete_dims.items():
+        t_values = torch.tensor(
+            values, device=unique_neighbors.device, dtype=unique_neighbors.dtype
+        )
+        idx = unique_neighbors[:, dim].long().clamp(0, len(values) - 1)
```

Comment:
Why do we need to clamp here? Both `plus_neighbors` and `minus_neighbors` have been clamped, right? If we actually need to clamp here, I think that may produce duplicate neighbors.

Reply:
I will investigate this, after fixing the clamping from above.

Reply:
Good catch, it was not necessary and is removed now.